### PR TITLE
full_dask_inclusion

### PR DESCRIPTION
**Pull request by MarjanAsgari**

### CODE REVIEW

**Comment by MarjanAsgari** [comment]
Hi @valhassan, 
I can't add you as a reviewer as it's grayed out. Tag you instead :)
The full project is just committed. It is a stand alone dask project, which can be run without any need to geo-deep-learning. 
There are two main scripts which are geo_inference_dask.py and geo_dask.py.

File: geo_inference/config/sample.yaml
Comment (valhassan) [inline geo_inference/config/sample.yaml:1]: The paths in this file should be left bland or generic.

File: geo_inference/geo_dask.py [hunk 1]
```diff
@@ -0,0 +1,1136 @@
...
+            )
+            .cpu()
+            .numpy()[0]
+        )
+        ready_np_array = (ready_np_array * 255).astype(np.int32)
+        return ready_np_array[np.newaxis, ...]
+    return aoi_chunk  # If the chunk size is 0, return the original chunk
+
+
+def minmax_scale(
```
Comment (valhassan) [inline geo_inference/geo_dask.py:396]: Scaling is scripted in the models; is this function used elsewhere for something else?

File: geo_inference/geo_dask.py [hunk 2]
```diff
@@ -0,0 +1,1136 @@
...
+                    axis=0,
+                )
+        except Exception as e:
+            logger.error(f"Error occured in IrunModel: {e}")
+        finally:
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()  # Release unused memory
+
+
+def runModel_partial_neighbor(
```
Comment (valhassan) [inline geo_inference/geo_dask.py:705]: If runModel_full_neighbor provides the best performance, this function is unnecessary.

File: geo_inference/geo_dask.py [hunk 3]
```diff
@@ -0,0 +1,1136 @@
...
+if str(Path(__file__).parents[0]) not in sys.path:
+    sys.path.insert(0, str(Path(__file__).parents[0]))
+
+from utils.helpers import get_device
+
+
+logger = logging.getLogger(__name__)
+
+
+class DaskAOI(object):
```
[the remark this answers is not in the document]
Comment (MarjanAsgari) [reply geo_inference/geo_dask.py:36]: I agree, I just added here for now to have a stand-alone app to run. But based on the agreement, I can definitely move it to geo-deep-learning.

File: geo_inference/geo_dask.py (hunk 1 above)
Comment (MarjanAsgari) [reply geo_inference/geo_dask.py:396]: It was used in AOI, when reading tiff files and and right before running enhancement on them. So, this function is only being used to make data ready for enhancement.

File: geo_inference/geo_dask.py (hunk 2 above)
Comment (MarjanAsgari) [reply geo_inference/geo_dask.py:705]: Correct! I created an issue on this to ultimately remove it!

File: geo_inference/geo_dask.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+import torch
+import logging
+
+import numpy as np
+import scipy.signal.windows as w
+
+from pathlib import Path
+from scipy.special import expit
+
+if str(Path(__file__).parents[0]) not in sys.path:
```
Comment (valhassan) [inline geo_inference/geo_dask.py:11]: I don't think directly appending the code path to system path is good practice in this way; can we avoid this?

File: geo_inference/geo_inference_dask.py
Comment (mpelchat04) [inline geo_inference/geo_inference_dask.py:1]: Is the intent to replace geo_inference.py with geo_inference_dask.py? If so, the README should be modified accordingly.   
But I personnally suggest to rename geo_inference_dask.py with geo_inference.py.

File: geo_inference/geo_inference_dask.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+                }
+                all_bands_requested = []
+                
+                with rasterio.Env(**rio_gdal_options):
+                    for key, value in bands_requested.items():
+                        all_bands_requested.append(rioxarray.open_rasterio(value["meta"].href, chunks=stride_patch_size))
+                self.aoi_dask_array = xr.concat(all_bands_requested, dim="band")
+                del all_bands_requested
+
+            if bbox != "None":
```
Comment (mpelchat04) [inline geo_inference/geo_inference_dask.py:233]: should be:  
`if bbox is not None:`

File: geo_inference/geo_inference_dask.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                trim=False,
+                boundary="none",
+                dtype=np.uint8,
+            )
+            
+        
+            with ResourceProfiler(dt=1) as prof:
+                with ProgressBar() as pbar:
+                    pbar.register()
+                    import rioxarray
```
Comment (jfbourgon) [inline geo_inference/geo_inference_dask.py:297]: I suggest moving as global import

File: geo_inference/geo_inference_dask.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
+import os
+import sys
```
Comment (jfbourgon) [inline geo_inference/geo_inference_dask.py:2]: Since imports from relative path were remove, the sys module is no longer required

File: geo_inference/geo_inference_dask.py [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+            except asyncio.CancelledError:
+                logger.info("The End of Inference")
+        
+        asyncio.run(run_async())
+
+    async def async_run_inference(self,
+        inference_input: Union[Path, str],
+        bands_requested: List[str],
+        patch_size: int = 1024,
+        num_workers: int = 8,
```
Comment (jfbourgon) [inline geo_inference/geo_inference_dask.py:134]: To be more generic, I suggest relying on CPU count as the default instead of an hardcoded default value

File: geo_inference/geo_inference_dask.py (hunk 6 above)
Comment (MarjanAsgari) [reply geo_inference/geo_inference_dask.py:297]: Unfortunately about   rioxarray it should be local, as if I import it global, it doesn't recognize it.

File: README.md [hunk 9]
```diff
@@ -36,18 +36,33 @@ python geo_inference -a <args>
 ```
 - `-a`, `--args`: Path to arguments stored in yaml, consult ./config/sample_config.yaml
 ```bash
-python geo_inference -i <image> -m <model> -wd <work_dir> -bs <batch_size> -v <vec> -d <device> -id <gpu_id>
+python geo_inference -i <image> -br <bands_requested> -m <model> -wd <work_dir> -bs <batch_size> -v <vec> -d <device> -id <gpu_id> -cls <classes> -mg <mgpu>
 ```
 - `-i`, `--image`: Path to Geotiff
 - `-bb`, `--bbox`: AOI bbox in this format "minx, miny, maxx, maxy" (Optional)
+- `-br`, `--bands_requested`: The requested bands from provided Geotiff (if not provided, it uses all bands)
+<<<<<<< HEAD
```
Comment (mpelchat04) [inline README.md:44]: Merge conflicts not resolved here.

Comment (MarjanAsgari) [reply README.md]: I believe it should be good now

File: README.md [hunk 10]
```diff
@@ -36,18 +36,22 @@ python geo_inference -a <args>
...
-python geo_inference -i <image> -m <model> -wd <work_dir> -bs <batch_size> -v <vec> -d <device> -id <gpu_id>
+python geo_inference -i <image> -br <bands_requested> -m <model> -wd <work_dir> -bs <batch_size> -v <vec> -d <device> -id <gpu_id> -cls <classes> -mg <mgpu>
 ```
 - `-i`, `--image`: Path to Geotiff
 - `-bb`, `--bbox`: AOI bbox in this format "minx, miny, maxx, maxy" (Optional)
+- `-br`, `--bands_requested`: The requested bands from provided Geotiff (if not provided, it uses all bands)
 - `-m`, `--model`: Path or URL to the model file
 - `-wd`, `--work_dir`: Working Directory
-- `-bs`, `--batch_size`: The Batch Size
+- `-bs`, `--batch_size`: The Batch Size, the size of dask chunks, Default = 1024
```
Comment (mpelchat04) [inline README.md:46]: in helpers.py, it's -bs --patch_size for this parameter. Maybe normalize to -ps --patch_size.

Comment (MarjanAsgari) [reply README.md:46]: done, will wait for more comments, then will push it to the repo!
